### PR TITLE
Scale chroma distortion.

### PR DESCRIPTION
**Pull request by tdaede**
@barrbrain can you rerun your tuning on top of this patch

### CODE REVIEW

**Comment by barrbrain** [comment]
> @barrbrain can you rerun your tuning on top of this patch

@tdaede Thank you for this. I will start another collection of stats.

**Comment by barrbrain** [comment]
I adapted my [Configurable chroma dc_qi and ac_qi interpolation](https://github.com/barrbrain/rav1e/commit/71cf62c369719468cb89728aaec3efc8ef883bf9) patch and collected for subset3 at speed preset 4.

Repeating the previous approach and iterating over subsets to find a confidence interval: 0.316±0.001 or 81 in Q8, essentially the original value from Daala.

chroma-balance

**Comment by barrbrain** [comment]
I tested [this combination on AWCY](https://beta.arewecompressedyet.com/?job=master-79252e6a0cc76964d9d4a7de266cfaaad7804f54&job=chroma-dist-tuned%402019-08-24T09%3A29%3A26.194Z) and the result was not great:

|   PSNR | PSNR Cb | PSNR Cr | PSNR HVS |   SSIM | MS SSIM | CIEDE 2000 |
|   ---: |    ---: |    ---: |     ---: |   ---: |    ---: |       ---: |
| 1.9315 |  7.2280 |  8.6807 |   2.0982 | 1.5924 |  1.8361 |     5.1297 |

I also tested [with --tune Psnr on AWCY](https://beta.arewecompressedyet.com/?job=master-79252e6a-psnr%402019-08-24T10%3A43%3A21.717Z&job=chroma-dist-tuned-psnr%402019-08-24T10%3A42%3A35.315Z) which was not any better:

|   PSNR | PSNR Cb | PSNR Cr | PSNR HVS |   SSIM | MS SSIM | CIEDE 2000 |
|   ---: |    ---: |    ---: |     ---: |   ---: |    ---: |       ---: |
| 3.2891 |     N/A | 17.9523 |   4.2194 | 2.7868 |  3.4516 |     7.9285 |

**Comment by barrbrain** [comment]
For reference, [intra-only tells a different story](https://beta.arewecompressedyet.com/?job=master-79252e6a-psnr-subset1%402019-08-24T12%3A52%3A08.992Z&job=chroma-dist-tuned-psnr-subset1%402019-08-24T12%3A51%3A32.057Z):

|   PSNR | PSNR Cb |  PSNR Cr | PSNR HVS |   SSIM | MS SSIM | CIEDE 2000 |
|   ---: |    ---: |     ---: |     ---: |   ---: |    ---: |       ---: |
| 3.0248 | -8.2096 | -13.2757 |   3.1171 | 3.1988 |  3.1970 |     0.6644 |

File: src/rate.rs [hunk 1]
```diff
@@ -573,8 +574,16 @@ impl QuantizerParameters {
     let scale = q57(QSCALE + bit_depth as i32 - 8);
     let quantizer = bexp64(log_target_q + scale);
     let (offset_u, offset_v) = chroma_offset(log_target_q);
-    let quantizer_u = bexp64(log_target_q + offset_u + scale);
-    let quantizer_v = bexp64(log_target_q + offset_v + scale);
+    let log_target_q_u = log_target_q + offset_u + scale;
+    let log_target_q_v = log_target_q + offset_v + scale;
+    let quantizer_u = bexp64(log_target_q_u);
+    let quantizer_v = bexp64(log_target_q_v);
```
Comment (barrbrain) [inline src/rate.rs]: These values are subtly different from what we expect. For correct `lambda_u` and `lambda_v`, rather:
```rust
let log_target_q_u = log_target_q + offset_u;
let log_target_q_v = log_target_q + offset_v;
let quantizer_u = bexp64(log_target_q_u + scale);
let quantizer_v = bexp64(log_target_q_v + scale);
```

**Comment by barrbrain** [comment]
I repeated the collection and analysis with the fix for `scale`: 0.290±0.001 or 74 in Q8:

chroma-balance

Results for [subset1 on AWCY](https://beta.arewecompressedyet.com/?job=master-79252e6a-psnr-subset1%402019-08-24T12%3A52%3A08.992Z&job=chroma-dist-tuning-psnr-subset1%402019-08-25T15%3A59%3A37.986Z) seem reasonable:

|   PSNR |  PSNR Cb |  PSNR Cr | PSNR HVS |   SSIM | MS SSIM | CIEDE 2000 |
|   ---: |     ---: |     ---: |     ---: |   ---: |    ---: |       ---: |
| 2.2579 | -22.1874 | -25.1877 |   2.3843 | 2.2387 |  2.3291 |    -3.4330 |

Results for [objective-1-fast on AWCY](https://beta.arewecompressedyet.com/?job=master-79252e6a-psnr%402019-08-24T10%3A43%3A21.717Z&job=chroma-dist-fixed-psnr%402019-08-25T21%3A00%3A37.717Z) are terrible for some sequences:

|   PSNR | PSNR Cb |  PSNR Cr | PSNR HVS |   SSIM | MS SSIM | CIEDE 2000 |
|   ---: |    ---: |     ---: |     ---: |   ---: |    ---: |       ---: |
| 7.1934 | -8.0033 | -14.5493 |   6.7523 | 7.9560 |  7.5724 |     0.8623 |

File: src/encoder.rs [hunk 2]
```diff
@@ -1208,6 +1211,7 @@ pub fn encode_tx_block<T: Pixel>(
     let tx_dist_scale_bits = 2 * (3 - get_log_tx_scale(tx_size));
     let tx_dist_scale_rounding_offset = 1 << (tx_dist_scale_bits - 1);
     tx_dist = (tx_dist + tx_dist_scale_rounding_offset) >> tx_dist_scale_bits;
+    tx_dist = (fi.dist_scale[0] * tx_dist as f64) as i64
```
Comment (barrbrain) [inline src/encoder.rs]: This function processes a single plane. Perhaps:
```suggestion
    tx_dist = (fi.dist_scale[p] * tx_dist as f64) as i64
```

File: src/encoder.rs [hunk 3]
```diff
@@ -1879,7 +1884,7 @@ pub fn write_tx_blocks<T: Pixel>(
               || skip
               || dist >= 0
           );
-          tx_dist += dist;
+          tx_dist += ((dist as f64) * fi.dist_scale[p]) as i64;
```
Comment (barrbrain) [inline src/encoder.rs]: This same change also needs to be made to `write_tx_tree()`.

File: src/encoder.rs [hunk 4]
```diff
@@ -1208,6 +1211,7 @@ pub fn encode_tx_block<T: Pixel>(
     let tx_dist_scale_bits = 2 * (3 - get_log_tx_scale(tx_size));
     let tx_dist_scale_rounding_offset = 1 << (tx_dist_scale_bits - 1);
     tx_dist = (tx_dist + tx_dist_scale_rounding_offset) >> tx_dist_scale_bits;
+    tx_dist = (fi.dist_scale[p] * tx_dist as f64) as i64
```
Comment (barrbrain) [inline src/encoder.rs]: This change doubles up with the changes in `write_tx_blocks()`.

File: src/rate.rs [hunk 5]
```diff
@@ -572,9 +573,17 @@ impl QuantizerParameters {
   ) -> QuantizerParameters {
     let scale = q57(QSCALE + bit_depth as i32 - 8);
     let quantizer = bexp64(log_target_q + scale);
-    let (offset_u, offset_v) = chroma_offset(log_target_q);
-    let quantizer_u = bexp64(log_target_q + offset_u + scale);
-    let quantizer_v = bexp64(log_target_q + offset_v + scale);
+    let (offset_u, offset_v) = chroma_offset(log_target_q, uv_m_q8);
```
Comment (barrbrain) [inline src/rate.rs]: It looks like `uv_m_q8` slipped in by a rebase error.
```suggestion
    let (offset_u, offset_v) = chroma_offset(log_target_q);
```

File: src/encoder.rs [hunk 6]
```diff
@@ -2017,7 +2021,7 @@ pub fn write_tx_tree<T: Pixel>(
               || skip
               || dist >= 0
           );
-          tx_dist += dist;
+          tx_dist += ((dist as f64) * fi.dist_scale[p]) as i64;
```
Comment (ycho) [inline src/encoder.rs]: Ok, so luma and chroma distortions can be scaled in tx-domain as well.

File: src/rate.rs [hunk 7]
```diff
@@ -589,8 +598,8 @@ impl QuantizerParameters {
         select_ac_qi(quantizer_u, bit_depth).max(1),
         select_ac_qi(quantizer_v, bit_depth).max(1),
       ],
-      lambda: (::std::f64::consts::LN_2 / 6.0)
-        * ((log_target_q as f64) * Q57_SQUARE_EXP_SCALE).exp(),
+      lambda,
+      dist_scale: [1.0, lambda / lambda_u, lambda / lambda_v],
```
Comment (ycho) [inline src/rate.rs]: Hmm, but scalers for YUV are derived from lambdas?! Isn't that essentially equivalent to applying different lambda values when computing rd_cost?

File: src/rdo.rs [hunk 8]
```diff
@@ -422,6 +425,8 @@ fn compute_tx_distortion<T: Pixel>(
     (tx_dist as f64 * bias) as u64
```
Comment (ycho) [inline src/rdo.rs]: I think 'tx_dist' here has been already scaled by fi.dist_scale[0], when it is passed from encode_block_post_cdef()?

File: src/rdo.rs [hunk 9]
```diff
@@ -422,6 +425,8 @@ fn compute_tx_distortion<T: Pixel>(
...
   };
 
+  distortion = (fi.dist_scale[0] * distortion as f64) as u64;
```
Comment (ycho) [inline src/rdo.rs]: Thus, if not skipped block, this seems wrong to me since distortion is already has been scaled.

**Comment by tdaede** [comment]
@ycho I have updated this PR with @barrbrain 's fixes, so it should now be correct. The only thing that I didn't include is the stats gathering and interpolation patch, as I think that should be a separate pr.

**Comment by ycho** [comment]
@tdaede , can you run this PR on awcy pls? so that we can see the image frames as I requested.

**Comment by tdaede** [comment]
@ycho https://beta.arewecompressedyet.com/?job=master-2019-10-09_182754-8b19a94&job=chroma_dist-2019-10-09_182802-d946584

**Comment by ycho** [comment]
> @ycho https://beta.arewecompressedyet.com/?job=master-2019-10-09_182754-8b19a94&job=chroma_dist-2019-10-09_182802-d946584

Thanks.
I've checked several images, and seeing mostly improvements with this PR increased bitstream size little bit.
For ex, what do you think for [Clovis](https://beta.arewecompressedyet.com/analyzer/?maxFrames=4&decoder=https://people.xiph.org/~mbebenita/analyzer/inspect.js&decoderName=master-2019-10-09_182754-8b19a94&file=https://beta.arewecompressedyet.com/runs/master-2019-10-09_182754-8b19a94/subset1/Clovisfest.y4m-172.ivf&decoder=https://people.xiph.org/~mbebenita/analyzer/inspect.js&decoderName=chroma_dist-2019-10-09_182802-d946584&file=https://beta.arewecompressedyet.com/runs/chroma_dist-2019-10-09_182802-d946584/subset1/Clovisfest.y4m-172.ivf)?

**Comment by ycho** [comment]
> @ycho https://beta.arewecompressedyet.com/?job=master-2019-10-09_182754-8b19a94&job=chroma_dist-2019-10-09_182802-d946584
(1 frame only)

|   PSNR |  PSNR Cb |  PSNR Cr | PSNR HVS |   SSIM | MS SSIM | CIEDE 2000 |
|   ---: |     ---: |     ---: |     ---: |   ---: |    ---: |       ---: |
| 0.6297 | -12.1790 | -14.6372 |   0.8292 | 0.6245 |  0.7480 |    -2.2254 |

Apart from subjective quality gains, do we think -2.22 ciede2000 worth 0.62 psnr loss (1 frame only)?

**Comment by tdaede** [comment]
@ycho yes, not because I think this particular tradeoff is worth it, but because we can make a better tradeoff - like @barrbrain 's retuning on top of this patch.

**Comment by ycho** [comment]
> @ycho yes, not because I think this particular tradeoff is worth it, but because we can make a better tradeoff - like @barrbrain 's retuning on top of this patch.

Okay, then no objection on landing this PR but I'd like to re-remind us please check the visual quality with our eyes whenever we change subjective qualities.

**Comment by ycho** [comment]
[Also started awcy run for objecctive-1-fast](https://beta.arewecompressedyet.com/?job=master-2019-10-09_182754-8b19a94%402019-10-10T21%3A35%3A38.845Z&job=chroma_dist-2019-10-09_182802-d946584%402019-10-10T21%3A36%3A05.969Z)

**Comment by tdaede** [comment]
Yes, that's been the longstanding issue with this changeset. It suffers less after @barrbrain 's follow up patch, but still isn't great. I've basically accepted that it's random followout from the large change in chroma distortion - it's pushed lambda around in an unfavorable direction on this clip.

**Pull request merged**